cmd/helmhook/hook/addon.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright (C) 2022-2024 ApeCloud Co., Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hook
+
+import (
+	"context"
+	"encoding/json"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	errors "sigs.k8s.io/controller-runtime/pkg/client"
+
+	extensionsv1alpha1 "github.com/apecloud/kubeblocks/apis/extensions/v1alpha1"
+	"github.com/apecloud/kubeblocks/pkg/client/clientset/versioned"
+)
+
+type Addon struct {
+	BasedHandler
+
+	AddonEnabledUpdated bool
+}
+
+const (
+	helmResourcePolicyKey  = "helm.sh/resource-policy"
+	helmResourcePolicyKeep = "keep"
+)
+
+func (p *Addon) IsSkip(*UpgradeContext) (bool, error) {
+	return p.AddonEnabledUpdated, nil
+}
+
+func (p *Addon) Handle(ctx *UpgradeContext) (err error) {
+	addons, err := ctx.KBClient.ExtensionsV1alpha1().Addons().List(ctx, metav1.ListOptions{
+		LabelSelector: toLabelSelector(addonSelectorLabels()),
+	})
+
+	if err != nil {
+		return errors.IgnoreNotFound(err)
+	}
+
+	for _, addon := range addons.Items {
+		if addon.GetDeletionTimestamp() != nil {
+			continue
+		}
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			return patchDisableUpdateAddon(ctx, ctx.KBClient, addon)
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func patchDisableUpdateAddon(ctx context.Context, client *versioned.Clientset, addon extensionsv1alpha1.Addon) error {
+	annotations := addon.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	if addon.GetAnnotations()[helmResourcePolicyKey] == helmResourcePolicyKeep {
+		return nil
+	}
+	annotations[helmResourcePolicyKey] = helmResourcePolicyKeep
```

Comment:
addon.GetAnnotations()[helmResourcePolicyKey] == helmResourcePolicyKeep should be true arriving here.
why set it again?